File: app/src/androidTest/java/ru/org/adons/cblock/data/BlockListModelTest.java
package ru.org.adons.cblock.data;

import android.text.format.DateUtils;

import org.junit.Before;
import org.junit.Test;

import java.util.Date;

import javax.inject.Inject;

import ru.org.adons.cblock.model.BlockListItem;
import ru.org.adons.cblock.model.CallLogItem;
import ru.org.adons.cblock.utils.Logging;
import rx.Observable;
import rx.observers.TestSubscriber;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;

/**
 * Android test for {@link BlockListModel}
 */

public class BlockListModelTest extends BaseModelTest {

    @Inject BlockListModel blockListModel;
    private final String phoneNumber1 = "123456789";
    private final String phoneNumber2 = "987654321";

    @Before
    public void setUp() {
        // provide dependencies
        testComponent().inject(this);
        // clear all local data before test
        blockListModel.clearBlockList();
    }

    @Test
    public void addNumberToBlockList() {
        Logging.d("check number added to block list");
        addNumber(phoneNumber1);
        //
        Logging.d("check not duplicated");
        addNumber(phoneNumber1);
        //
        Logging.d("check add other number");
        addOtherNumber();
        //
        Logging.d("check not duplicated");
        addSameTwoNumbers();
    }

    private void addNumber(String phoneNumber) {
        TestSubscriber<BlockListItem> testSubscriber = new TestSubscriber<>();
        blockListModel.addNumber(getLogItem(phoneNumber));
        blockListModel.getBlockList()
                .flatMap(Observable::from)
                .map(listItem -> {
                    assertNotNull(listItem);
                    assertEquals(listItem.getPhoneNumber(), phoneNumber);
                    return printItemDetails(listItem);
                })
                .subscribe(testSubscriber);
        testSubscriber.assertValueCount(1);
        testSubscriber.assertNoErrors();
        testSubscriber.assertCompleted();
        testSubscriber.assertUnsubscribed();
    }

    private void addOtherNumber() {
        TestSubscriber<BlockListItem> testSubscriber = new TestSubscriber<>();
        blockListModel.addNumber(getLogItem(phoneNumber2));
        blockListModel.getBlockList()
                .map(blockList -> {
                    BlockListItem listItem = blockList.get(0);
                    BlockListItem listItem1 = blockList.get(1);
                    assertFalse(listItem.getPhoneNumber().equals(listItem1.getPhoneNumber()));
                    return blockList;
                })
                .flatMap(Observable::from)
                .map(this::printItemDetails)
                .subscribe(testSubscriber);
        testSubscriber.assertValueCount(2);
        testSubscriber.assertNoErrors();
        testSubscriber.assertCompleted();
        testSubscriber.assertUnsubscribed();
    }

    private void addSameTwoNumbers() {
        TestSubscriber<BlockListItem> testSubscriber = new TestSubscriber<>();
        blockListModel.addNumber(getLogItem(phoneNumber1));
        blockListModel.addNumber(getLogItem(phoneNumber2));
        blockListModel.getBlockList()
                .flatMap(Observable::from)
                .map(this::printItemDetails)
                .subscribe(testSubscriber);
        testSubscriber.assertValueCount(2);
        testSubscriber.assertNoErrors();
        testSubscriber.assertCompleted();
        testSubscriber.assertUnsubscribed();
    }

    private CallLogItem getLogItem(String phoneNumber) {
        return CallLogItem.builder()
                .setId(1L)
                .setPhoneNumber(phoneNumber)
                .setDate(new Date().getTime())
                .build();
    }

    private BlockListItem printItemDetails(BlockListItem item) {
        Logging.d("id=" + item.getId().toString());
        Logging.d("phone=" + item.getPhoneNumber());
        Logging.d("date=" + DateUtils.getRelativeTimeSpanString(item.getDate()).toString());
        Logging.d("name=" + item.getName());
        return item;
    }

}

File: app/src/androidTest/java/ru/org/adons/cblock/data/BlockListModelTest.kt
package ru.org.adons.cblock.data

import android.text.format.DateUtils
import junit.framework.Assert.*
import org.junit.Before
import org.junit.Test
import ru.org.adons.cblock.ext.log
import ru.org.adons.cblock.model.BlockListModel
import rx.Observable
import rx.observers.TestSubscriber
import java.util.*
import javax.inject.Inject

/**
 * Android test for [BlockListModel]
 */
class BlockListModelTest : BaseModelTest() {

    private val phoneNumber1 = "+380123456789"
    private val phoneNumber2 = "+380987654321"
    private val name1 = "Test Name"

    @Inject lateinit var blockListModel: BlockListModel

    @Before
    fun setUp() {
        // provide dependencies
        testComponent().inject(this)
        // clear all local data before test
        blockListModel.clearBlockList()
    }

    @Test
    fun addNumberToBlockList() {
        log("check number added to block list")
        addNumber(phoneNumber1)
        //
        log("check not duplicated")
        addNumber(phoneNumber1)
        //
        log("check add other number")
        addOtherNumber()
        //
        log("check not duplicated")
        addSameTwoNumbers()
    }

    private fun addNumber(phoneNumber: String) {
        val testSubscriber = TestSubscriber<BlockListItem>()
        blockListModel.addNumber(getLogItem(phoneNumber, name1))
        blockListModel.getBlockList()
                .flatMap { Observable.from(it) }
                .doOnNext {
                    assertNotNull(it)
                    assertEquals(it.phoneNumber, phoneNumber)
                    printItemDetails(it)
                }
                .subscribe(testSubscriber)
        testSubscriber.assertValueCount(1)
        testSubscriber.assertNoErrors()
        testSubscriber.assertCompleted()
        testSubscriber.assertUnsubscribed()
    }

    private fun addOtherNumber() {
        val testSubscriber = TestSubscriber<BlockListItem>()
        blockListModel.addNumber(getLogItem(phoneNumber2, null))
        blockListModel.getBlockList()
                .doOnNext {
                    val listItem = it[0]
                    val listItem1 = it[1]
                    assertFalse(listItem.phoneNumber == listItem1.phoneNumber)
                }
                .flatMap { Observable.from(it) }
                .doOnNext { this.printItemDetails(it) }
                .subscribe(testSubscriber)
        testSubscriber.assertValueCount(2)
        testSubscriber.assertNoErrors()
        testSubscriber.assertCompleted()
        testSubscriber.assertUnsubscribed()
    }

    private fun addSameTwoNumbers() {
        val testSubscriber = TestSubscriber<BlockListItem>()
        blockListModel.addNumber(getLogItem(phoneNumber1, name1))
        blockListModel.addNumber(getLogItem(phoneNumber2, null))
        blockListModel.getBlockList()
                .flatMap { Observable.from(it) }
                .doOnNext { this.printItemDetails(it) }
                .subscribe(testSubscriber)
        testSubscriber.assertValueCount(2)
        testSubscriber.assertNoErrors()
        testSubscriber.assertCompleted()
        testSubscriber.assertUnsubscribed()
    }

    private fun getLogItem(phoneNumber: String, name: String?): CallLogItem = CallLogItem(
            1L, phoneNumber, Date().time, name, false, false)

    private fun printItemDetails(item: BlockListItem): BlockListItem = with(item) {
        log("id=" + id.toString())
        log("phone=" + phoneNumber)
        log("date=" + DateUtils.getRelativeTimeSpanString(date))
        log("name=" + name)
        item
    }

}
